Task: Add Todo App

christianaguilargithub opened the pull request:
horray! :)

christianaguilargithub committed:
main code for review

jorenrui commented on .gitignore:
```diff
@@ -0,0 +1,23 @@
+.env.test.local
+.env.production.local
+
+npm-debug.log*
```
I think you should add `.eslintcache` and `debug.log` here, then remove these files in git?

christianaguilargithub replied:
done . no errors

jorenrui replied:
I think you forgot to `git push` the changes here?

christianaguilargithub replied:
pushed it in another branch named enhancement  should I make a new PR? but not all are finished . I'm doing one commit per changes

jorenrui replied:
Ahh... all the requested changes in this branch should be done on this branch then pushed it here. Otherwise, the code reviewer wouldn't be able to see the changes you've made and reviewed it again.

christianaguilargithub replied:
can I pushed the changes don on a seperate branch to this branch or should I check out to this branch and redo all the changes?

jorenrui replied:
You can do:

// Go to `feature-todo`
git checkout feature-todo

// Get changes from the other branch named `enhancements`?
git merge enhancements

// Push changes to GitHub
git push

christianaguilargithub replied:
Awesome! done merging

jorenrui commented on src/App.js:
```diff
@@ -0,0 +1,64 @@
+  const [filteredTodos, setFilteredTodos] = useState([]);
+  //Run once when the app start
+  useEffect(
+    ()=>{
```
For code style consistency, you can setup eslint. Tho that change should be done in a separate PR branching out from this branch. I'll create an issue card for that in a bit.

christianaguilargithub replied:
Dons installing EsLint , changed code  style using Eslint show fixes and converting arrow functions to named functions . Am I right? Is it the consistency of style?

jorenrui replied:
Hmm... I was kinda talking about the whitespace like tab, lines, spaces. For useEffect like this, it's best to use arrow functions.

jorenrui replied:
Also I think you didn't apply the changes here in `feature-todo` and `git push` it so I wasn't able to see the changes you made?

christianaguilargithub replied:
ESlint is awesome for correcting and keeping track of code changes :)

jorenrui replied:
I think it's best to finish first the requested changes in this PR here before working on a new issue in a new PR?

christianaguilargithub replied:
affirmative :)

jorenrui commented on src/App.js:
```diff
@@ -0,0 +1,64 @@
+  const saveLocalTodos = () => {
+      localStorage.setItem("todos", JSON.stringify(todos));
+  };
+  const getLocalTodos = () => {
```
I'm assuming this is kinda like your APIs (fetchers). For those kinds of files, it is recommended to put it in a separate folder like `utils/api` then import it in here. Since in the future, one may possible reuse those APIs in another component.

christianaguilargithub replied:
created New folder named utils inside a folder named api / todos.js , will configure how to transfer api(fetching codes) codes and connect it with no errors.

jorenrui commented on src/App.js:
```diff
@@ -0,0 +1,64 @@
+    },[todos,status]);
+  
+  //Save to Local
+  const saveLocalTodos = () => {
```
Same as above. Recommend relocating this to `utils/api`. Then add `todos` as a parameter.

jorenrui replied:
Also I think this code isn't used anywhere or am I wrong?

jorenrui commented on src/App.js:
```diff
@@ -0,0 +1,64 @@
+      localStorage.setItem("todos", JSON.stringify(todos));
+  };
+  const getLocalTodos = () => {
+    if(localStorage.getItem("todos")=== null) {
```
Redundant usage of `localStorage.getItem("todos")`? You should store this in a variable like:

```js
const todos = localStorage.getItem("todos");

// No need for strict equality so that you can also catch "undefined" values 
if (todos == null) { // <val> === <val> -> <val> == <val>
  ...
} else {
  setTodos(JSON.parse(todos));
}
```

jorenrui commented on src/App.js:
```diff
@@ -0,0 +1,64 @@
+    if(localStorage.getItem("todos")=== null) {
+      localStorage.setItem("todos", JSON.stringify([]));
+    } else {
+      let todoLocal = JSON.parse(localStorage.getItem("todos"));
```
Preferrable use `const` when there's no reassignment to a variable. Tho would recommend doing the code mentioned above like:

```js
setTodos(JSON.parse(todos));
```

jorenrui commented on src/App.js:
```diff
@@ -0,0 +1,64 @@
+      localStorage.setItem("todos", JSON.stringify([]));
+    } else {
+      let todoLocal = JSON.parse(localStorage.getItem("todos"));
+      setTodos(todoLocal);
```
Since `getLocalTodos` is kinda like your API. you should return the values here instead like:

```js
// `utils/api/todos.js`
export const getLocalTodos = () => {
  ...

  return todos;
}
```

Then in your `useEffect` do:

```js
useEffect(()=>{
  const todos = getLocalTodos();

  setTodos(todos);
}, []);
```

jorenrui commented on src/App.js:
```diff
@@ -0,0 +1,64 @@
+
+function App() {
+ 
+  //state stuffs
```
No need to comment this? Since it kinda clutters the code.

jorenrui commented on src/App.js:
```diff
@@ -0,0 +1,64 @@
+  //state stuffs
+  const[inputText, setInputText]= useState("");
+  const[todos, setTodos] = useState([]);
+  const [status, setStatus] = useState("all");
```
Preferred to create enums (tho this is only available in TypeScript) for this so that the limited choices (all, completed or uncompleted) is clear. You can either mock enums in JS like [here](https://stackoverflow.com/questions/287903/what-is-the-preferred-syntax-for-defining-enums-in-javascript)

or you can create constants to be put in `utils/constants/todo.js` like

```js
export const TodosFilter = {
  All: 'all',
  Completed: 'completed',
  Uncompleted: 'uncompleted',
} 
```

then set the value in state like:

```js
const [status, setState] = useState(TodosFilter.All)
```

Also how bout `filter` instead of `status`?

jorenrui commented on src/App.js:
```diff
@@ -0,0 +1,64 @@
+import React, {useState, useEffect} from "react";
+import './App.css';
+//importing components
```
For imports, how bout trying this format? Also no need to add comments unless you want to use the "ANCHOR Comments" extension of VSCode
```
import packages;
import packages;
import packages;

import local files;
import local files;
import local files;

import css;

jorenrui commented on src/App.js:
```diff
@@ -0,0 +1,64 @@
+          setFilteredTodos(todos);
+        } 
+      };
+      filterHandler();
```
I think it is unnecessary to wrap this in a function?

jorenrui commented on src/App.js:
```diff
@@ -0,0 +1,64 @@
+   useEffect(
+    () => {
+      //Functions and events
+      const filterHandler = ()=>{
```
Why not make `filteredTodos` a variable instead of a react state them put it outside of `useEffect`? Cause the filterHandler gets run when `todos` changes.

Since `todos` is a React state, any change of it will trigger re-renders.

So you can safely do:

```js
const [todos, setTodos] = useState([]);

const filteredTodos = status === 'all'
  ? todos
  : todos.filter((todo) => {
    if (status === 'completed') {
      return todo.completed;
    }

    if (status === 'uncompleted') {
      return todo.completed === false;
    }
  })
```

christianaguilargithub replied:
hindi ko magets paano to

jorenrui commented on src/App.js:
```diff
@@ -0,0 +1,64 @@
+      const filterHandler = ()=>{
+        switch (status){
+          case 'completed':
+            setFilteredTodos(todos.filter(todo => todo.completed === true));
```
Unnecessary checking of truthy value. You can do `todo.completed` instead of `todo.completed === true`.

jorenrui commented on src/App.js:
```diff
@@ -0,0 +1,64 @@
+
+  return (
+    <div className="App">
+      <header>
```
Since you're using a `header` tag, you can also wrap the next sibling of this with `main`.

jorenrui commented on src/App.js:
```diff
@@ -0,0 +1,64 @@
+      <header>
+        <h1>To Do List </h1>
+      </header>
+      <Form inputText={inputText} todos={todos} setTodos={setTodos} setInputText={setInputText} setStatus={setStatus} />
```
Preferred format:
```js
<Form
  inputText={inputText}
  todos={todos}
  setTodos={setTodos}
  ...
/>
```

jorenrui commented on src/App.js:
```diff
@@ -0,0 +1,64 @@
+        <h1>To Do List </h1>
+      </header>
+      <Form inputText={inputText} todos={todos} setTodos={setTodos} setInputText={setInputText} setStatus={setStatus} />
+      <TodoList filteredTodos={filteredTodos} setTodos={setTodos} todos={todos}/>
```
Same comment as above.

jorenrui commented on src/components/Todo.js:
```diff
@@ -0,0 +1,29 @@
+    return(
+        <div className="todo">
+            <li className={`todo-item ${todo.completed ? "completed" : ""}`}>{text}</li>
+            <button onClick={completeHandler} className="complete-btn"><i className="fas fa-check"></i></button>
```
For components/tags that has no children, you can make it a self closing tag like:

```js
<i className="fas fa-check" />
```

jorenrui commented on src/components/Todo.js:
```diff
@@ -0,0 +1,29 @@
+        <div className="todo">
+            <li className={`todo-item ${todo.completed ? "completed" : ""}`}>{text}</li>
+            <button onClick={completeHandler} className="complete-btn"><i className="fas fa-check"></i></button>
+            <button onClick={deleteHandler} className="trash-btn"><i className="fas fa-trash"></i></button>
```
Always add `type="button"` or `type="submit"` to buttons to avoid unexpected bugs in the future. Since when button is relocated inside a form, having no `type="button"` will trigger a "submit" action in which you might not expect.

jorenrui commented on src/components/Todo.js:
```diff
@@ -0,0 +1,29 @@
+    };
+   
+    const completeHandler = () => {
+        setTodos(todos.map(item =>{
```
How 'bout only checking `completed` since it's the only changing value:

```js
todos.map((item) => ({
  ...item,
  completed: item.id === todo.id
    ? !item.completed
    : item.completed
}))
```

Also prefer `todo` rather than `item`

jorenrui commented on src/components/Todo.js:
```diff
@@ -0,0 +1,29 @@
+const Todo = ({text,todo, todos,setTodos}) => {
+    //Events
+    const deleteHandler = () => {
+        setTodos(todos.filter((el) => el.id !==todo.id))
```
How 'bout not passing the `todos` prop and instead use it like this:

```js
setTodos((todos) => todos.filter(curTodo) => curTodo.id !== todo.id));
```

jorenrui commented on src/components/Form.js:
```diff
@@ -0,0 +1,38 @@
+import React from "react";
+
```
Unnecessary extra line?

jorenrui commented on src/components/Form.js:
```diff
@@ -0,0 +1,38 @@
+
+
+function Form({setStatus, setInputText, todos, setTodos, inputText}) {
+    
```
Remove extra line

jorenrui commented on src/components/Form.js:
```diff
@@ -0,0 +1,38 @@
+
+     const submitTodoHandler = (e) => {
+         e.preventDefault();
+         setTodos([
```
I think you should also check when `inputText` has a value to avoid adding an empty todo.

```js
if (inputText) {
  setTodos()
}
```

jorenrui commented on src/components/Form.js:
```diff
@@ -0,0 +1,38 @@
+     const submitTodoHandler = (e) => {
+         e.preventDefault();
+         setTodos([
+             ...todos, {text: inputText, completed: false, id: Math.random() }
```
Check first if the generated `Math.random()` has a duplicate value in the existing todo cause it might cause duplicate id bug.

jorenrui replied:
Also preferred to specify objects like this for readability purposes.

```js
{
  ...todos,
  {
    id: generatedId,
    text: inputText,
    completed: false,
  }
}
```

christianaguilargithub replied:
used Deno (web module) Import
import ShortUniqueId from 'https://cdn.jsdelivr.net/npm/short-unique-id@latest/short_uuid/mod.ts';
instead of Math.random(); for unique id's and no duplication

christianaguilargithub replied:
revert changes due to Error

jorenrui commented on src/components/Form.js:
```diff
@@ -0,0 +1,38 @@
+        <form>
+            <input value={inputText} onChange={inputTextHandler} type="text" className="todo-input" />
+            <button onClick={submitTodoHandler} className="todo-button" type="submit">
+            <i className="fas fa-plus-square"></i>
```
Can be a self closing tag

jorenrui commented on src/components/TodoList.js:
```diff
@@ -0,0 +1,21 @@
+const TodoList = ({ todos, setTodos, filteredTodos }) => {
+
+    
+    return (
```
For empty function blocks do this instead of explicitly returning:

```js
const TodoList = () => (
  <div>
    ...
  </div>
);
```

jorenrui commented on src/components/TodoList.js:
```diff
@@ -0,0 +1,21 @@
+           <ul className="todo-list">
+               {filteredTodos.map(todo =>(
+                   <Todo setTodos={setTodos} todos={todos} text={todo.text} key={todo.id} todo={todo} />
+
```
Remove extra line

jorenrui reviewed (changes requested):
Overall, I think you're all good naman in terms of React since you understand how hooks work. Just need to clean the code and styles na lang. That's why I suggest using eslint since it is possible that a company uses one if the project is done by a team. Tho you can do it in a separate PR na lang after fixing the changes here then merging it.

Also, I'll be giving a new task by adding an issue card in the GitHub repo

christianaguilargithub committed:
edited gitignore

christianaguilargithub committed:
removed saveTodos from App.js

christianaguilargithub committed:
Installed ESlint and React Router , edited app.js Form and TodoList for readability, edited TodoList from components Todo for readability

christianaguilargithub committed:
edited git ignore

jorenrui commented:
@christianaguilargithub just fix all the requested changes here. Also don't work on the [issues listed in here](https://github.com/christianaguilargithub/todo-review/issues) in this PR so that. You'll have to create a new PR for the ESLint setup and Post Page. But first you should finish the requested changes here and wait for this to be merged before working on those issues.

After all the requested changes has been done. You should click "Resolve Comment" on the ones you finished.

![image](https://user-images.githubusercontent.com/23741509/105658328-3ee91200-5f01-11eb-8fe3-0e3f8eebd367.png)

Then after resolving all the comments, you can re-request my review here:

![image](https://user-images.githubusercontent.com/23741509/105658169-f0d40e80-5f00-11eb-865f-27175a8665fe.png)

Then I'll review this Pr again

christianaguilargithub committed:
modifications gitignore modified:   src/components/Form.js modified:   src/components/Todo.js

christianaguilargithub committed:
arrange codes for consistency nothing changed except for spacings tabs for readability

christianaguilargithub committed:
fix lines renaming el => curTodo

christianaguilargithub committed:
revert changes

christianaguilargithub committed:
curTodo

christianaguilargithub committed:
fixing

christianaguilargithub committed:
still fixing

christianaguilargithub committed:
git problems resolve

christianaguilargithub committed:
insert and fix code in utils/api/todos.js

christianaguilargithub committed:
added export on const Todo function in components/Todo.js

christianaguilargithub committed:
fix todos.js in utils/api added return [];

christianaguilargithub committed:
TodoList.js explicitly returning resolved

christianaguilargithub committed:
added TodosFilter in todos.js , change status and setStatus in App.js to filter and setFilter

christianaguilargithub committed:
edited color in App.css

christianaguilargithub committed:
completeHander simplified setTodos task completed check

christianaguilargithub committed:
used Deno (web module) Import
import ShortUniqueId from 'https://cdn.jsdelivr.net/npm/short-unique-id@latest/short_uuid/mod.ts';
instead of Math.random(); for unique id's and no duplication

christianaguilargithub committed:
Revert changes in setTodos id because of errors

christianaguilargithub committed:
added Validation

christianaguilargithub committed:
recreated validation

christianaguilargithub committed:
added Date as ID

christianaguilargithub committed:
Solved blank entry and white space issues

christianaguilargithub committed:
white space entry only will not be accepted
